File: Book_图解算法/06_breadth-first_search/python/01_breadth-first_search.py
from collections import deque

def person_is_seller(name):
      return name[-1] == 'm'

graph = {}
graph["you"] = ["alice", "bob", "claire"]
graph["bob"] = ["anuj", "peggy"]
graph["alice"] = ["peggy"]
graph["claire"] = ["thom", "jonny"]
graph["anuj"] = []
graph["peggy"] = []
graph["thom"] = []
graph["jonny"] = []

def search(name):
    search_queue = deque()
    search_queue += [name]
    # This is how you keep track of which people you've searched before.
    searched = set()
    i = 0
    print(f"{i} = {search_queue}")
    while search_queue:
        i += 1
        person = search_queue.popleft()
        # Only search this person if you haven't already searched them.
        if person in searched:
            continue
        if person_is_seller(person):
            print(person + " is a mango seller!")
            return True
        search_queue += graph[person]
        print(f"{i}, search_queue = {search_queue}")
        # Marks this person as searched
        searched.add(person)
        print(f"{i}, searched = {searched}\n")
    return False

search("you")

# name = "you"
# search_queue = deque()
# search_queue += [name]
# # This is how you keep track of which people you've searched before.
# searched = set()
# while search_queue:
#     person = search_queue.popleft()
#     # Only search this person if you haven't already searched them.
#     if person in searched:
#         continue
#     if person_is_seller(person):
#         print(person + " is a mango seller!")
#         # return True
#     search_queue += graph[person]
#     # Marks this person as searched
#     searched.add(person)

File: 数据结构与算法/Book_图解算法/06_breadth-first_search/dart/01_breadth-first_search.dart
import 'dart:collection';

void main(List<String> args) {
  final graph = <String, List<String>>{};
  graph.addAll(
    <String, List<String>>{
      'you': ['alice', 'bob', 'claire'],
      'bob': ['anuj', 'peggy'],
      'alice': ['peggy'],
      'claire': ['thom', 'jonny'],
      'anuj': [],
      'peggy': [],
      'thom': [],
      'jonny': [],
    },
  );

  search(graph, 'you');
}

bool search(Map<String, List<String>> graph, String name) {
  final searchQueue = Queue()..addAll(graph[name] ?? []);
  final searched = List<String>.empty(growable: true);

  while (searchQueue.isNotEmpty) {
    final String person = searchQueue.removeFirst();
    if (searched.contains(person) == false) {
      if (_personIsSeller(person)) {
        print('$person is a Mango seller!');
        return true;
      } else {
        searchQueue.addAll(graph[person] ?? []);
        searched.add(person);
      }
    }
  }
  return false;
}

bool _personIsSeller(String name) {
  return name.endsWith('m');
}
